nixos/fossil-server: init

Uthar opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Fix #102220  

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Uthar committed:
nixos/fossil-server: init

aanderse commented on nixos/modules/services/development/fossil-server.nix:
```diff
@@ -0,0 +1,146 @@
+  config = mkIf cfg.enable {
+
+    networking.firewall.allowedTCPPorts = [ cfg.port ];
+    systemd.services.fossil = {
```
There isn't any real need to run this as `root`. Maybe `DynamicUser` would be a good choice instead.

Uthar replied:
Added the `user` and `group` options. I'm not sure if it's the best way, what do you think? The thing is that with customizable user/group, it's easy to manage permissions just by using `chmod`/`chown` and make `fossil` see _just_ what it needs to see. Also I'm not sure if `DynamicUser` would persist between reboots which is desirable.

aanderse commented on nixos/modules/services/development/fossil-server.nix:
```diff
@@ -0,0 +1,146 @@
+
+    networking.firewall.allowedTCPPorts = [ cfg.port ];
+    systemd.services.fossil = {
+      preStart = if lib.strings.hasSuffix ".fossil" cfg.repository
```
You should replace this directory creation with `systemd.tmpfiles.rules` entries instead.

Uthar replied:
You made me rethink this, and IMHO the place where `repository` points to should be created manually. Then, one can create/remove .fossil files as needed, and manage permissions etc.

aanderse commented on nixos/modules/services/development/fossil-server.nix:
```diff
@@ -0,0 +1,146 @@
+                 '';
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      script = "${pkgs.fossil}/bin/fossil server "
```
You can just use `ExecStart` instead.

Uthar replied:
Yeah, it's nice how it just embeds it in the `.service` file instead of creating a script in the store. Thanks.

aanderse commented on nixos/modules/services/development/fossil-server.nix:
```diff
@@ -0,0 +1,146 @@
+      description = "Do not compress HTTP replies";
+    };
+
+    nojail = mkOption {
```
From *quickly* reviewing documentation it seems like this option doesn't provide any real value when you're using `systemd`.

Uthar replied:
I just removed it, never used it before so would not know how to test it. If someone needs it in the future then it could be investigated.

aanderse commented on nixos/modules/services/development/fossil-server.nix:
```diff
@@ -0,0 +1,146 @@
+
+    jsmode = mkOption {
+      type = types.str;
+      default = "";
```
`types.enum` is what you're looking for here.

Uthar replied:
That's much cleaner, thanks - fixed

I noticed that the docstring does not keep newlines in `man configuration.nix`:
![image](https://user-images.githubusercontent.com/15697697/97781211-1499e980-1b8a-11eb-9ac5-f93fe5c8fe41.png)

Is this normal? How could I fix it?

aanderse reviewed (changes requested):
Thanks for your contribution. I have left a few comments I hope you find helpful. Additionally, it would be nice if the `systemd` unit had some hardening options applied against it.

If you have any questions at all, please do not hesitate to ask. I'm happy to help.

Uthar committed:
nixos/fossil-server: remove nojail option

Uthar committed:
nixos/fossil-server: change jsmode option type to enum

Uthar committed:
nixos/fossil-server: script->ExecStart

Uthar committed:
nixos/fossil-server: don't create directories automatically

User should manually create his Fossil repositories, then point the
Fossil server to them.

Uthar committed:
nixos/fossil-server: add user, group options

Allows to control what the Fossil server sees by setting chowns on fs,
and makes it persist between reboots

Uthar committed:
nixos/fossil-server: add systemd hardening options

Uthar commented on nixos/modules/services/development/fossil-server.nix:
```diff
@@ -149,6 +149,16 @@ in
         + (optionalString cfg.repolist "--repolist ")
         + "--port ${toString cfg.port} "
         + cfg.repository;
+        PrivateTmp = "yes";
+        NoNewPrivileges = true;
+        ProtectSystem = "strict";
+        CapabilityBoundingSet = "CAP_NET_BIND_SERVICE CAP_DAC_READ_SEARCH";
+        RestrictNamespaces = "uts ipc pid user cgroup";
+        ProtectKernelTunables = "yes";
+        ProtectKernelModules = "yes";
+        ProtectControlGroups = "yes";
+        PrivateDevices = "yes";
+        RestrictSUIDSGID = true;
```
Added some options which make `systemd-analyze security` say `fossil.service 4.9 OK`

Is there something to change?

aanderse commented:
Sorry I've been really busy. I'll try to get back to this at some point over the next few weeks.

Uthar committed:
nixos/fossil-server: set isSystemUser=true for cfg.user

Uthar committed:
nixos/fossil-server: stole + adjusted hardening config from nginx module

Uthar committed:
nixos/fossil-server: add cfg.repository to ReadWritePaths

Otherwise Fossil can't write to the repository file/repositories directory

Uthar committed:
nixos/fossil-server: add nixos/tests/fossil-server.nix

A test for clone, push and pull between server and client

Uthar commented:
Updated for 21.05 and added a test - please review! @aanderse  @SuperSandro2000

Uthar committed:
nixos/fossil-server: add localhost option

Uthar committed:
Merge branch 'master' into fossil-server

Uthar commented:
Resolved one merge conflict :)

Uthar committed:
oops, remove duplicate

srid commented:
Is there no interest in this?

nat-418 commented:
@etu what is the best way to revive this attempt to get a Fossil service in Nix?

SuperSandro2000 commented:
You can continue the work in another PR.
